### ISSUE TITLE
Failure to handle some modules on Windows due to encoding

### ISSUE THREAD

**Issue by adambenali** [seddonym/grimp#75]
Here is how to reproduce the error:

0. Use Windows
1. Install `grimp`.
2. Create a python package `test_encoding`. Inside `__init__.py` add this content : `print('變量')`.
3. Open a python shell and do:
```python
>>> import grimp
>>> graph = grimp.build_graph('test_encoding')
```
Fails with : 
```
  File "C:\Users\adam\Anaconda3\lib\site-packages\grimp\adaptors\filesystem.py", line 26, in read
    return file.read()
  File "C:\Users\adam\Anaconda3\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x8f in position 12: character maps to <undefined>
```

**Comment by seddonym** [comment]
Thanks for the bug report!

I'll look into it, though I don't have access to a Windows machine so I can't promise how easy it will be to fix. I'd welcome a pull request if you felt so inclined.

**Comment by seddonym** [comment]
Actually thinking about it, would you mind uploading the file you created here? I'll see if I can reproduce the problem with that file.

**Comment by adambenali** [comment]
Here you are : [test_encoding.zip](https://github.com/seddonym/grimp/files/4266766/test_encoding.zip).

**Comment by seddonym** [comment]
Thanks for that. Interestingly, that file loads fine for me. My guess is that what is happening is that your file is incorrectly encoded, and it's somehow being fixed in the process of transferring to me.

I think probably the right approach is for Grimp to handle file decoding problems in a more forgiving manner, but also you might want to look into why this is happening on your machine.

Incidentally, this is a good article on the subject. http://python-notes.curiousefficiency.org/en/latest/python3/text_file_processing.html

**Comment by adambenali** [comment]
Oh, that is curious. Maybe it depends on the region. What is the Windows version you are using ?
Also, could you please check the default encoding on the machine ? You can do that by typing `[System.Text.Encoding]::Default` in a PowerShell.

**Comment by seddonym** [comment]
I'm not using Windows. But my system's preferred encoding is UTF-8. What do you get if you run this?
```
python -c "import locale; print(locale.getpreferredencoding())"
```

**Comment by adambenali** [comment]
> I'm not using Windows.

This is precisely why it works for you.
As I implicitly mentioned before, this problem exists with Windows platforms (and more generally any platform that doesn't default its encoding to UTF-8). This is because the default encoding on Windows is not UTF-8, but rather [cp1252 & co](https://en.wikipedia.org/wiki/Windows_code_page) (which is the return value of `locale.getpreferredencoding()`).

Common solutions to this (rather classic) problem include : 
- Forcing the UTF-8 encoding
- Using a series of encoding guesses, starting with UTF-8, the most common one, and falling back to `latin-1` and the default platform encoding.
- Instead of `open`, using [`tokenize.open`](https://docs.python.org/3.4/library/tokenize.html#tokenize.open) which tries to detect file encodings
- Reading files in binary mode instead
- Using `open`'s _[errors](https://docs.python.org/3/howto/unicode.html#the-string-type)_ argument

**Comment by seddonym** [comment]
Hi - sorry for the delay. It feels to me that using `errors="backslashreplace"` is probably the right answer in this case. Would you agree? I'll try to put together a PR for you to review at some point (unless you beat me to it!)

**Comment by adambenali** [comment]
This would eliminate the exception but it would be replacing unrecognized characters with random characters. Although in the particular of `grimp` this is not essential (because we only care about imports, right ?), for me, using [`tokenize.open`](https://docs.python.org/3.4/library/tokenize.html#tokenize.open) would be better. We'd be preserving the integrity of the file as much as we can. What do you think ?

**Comment by seddonym** [comment]
Hi Adam - I've addressed this in PR https://github.com/seddonym/grimp/pull/76. Unfortunately it's not working with the coverage report at the moment, which I haven't had time to look into. If you had time to check that branch fixes the issue for you locally, that would be helpful!

**Comment by seddonym** [comment]
I've now released this in 1.2.1. Thanks for your help!
